[Threat Intel] Threat Intel Downloader implementation

chunyong-lin opened the pull request:
to: @jacknagz (terraform module, streamalert CLI) @ryandeivert (Lambda function code in `stream_alert/threat_intel_downloader/`)
cc: @airbnb/streamalert-maintainers
size: large
(Please don't review commits individually, I fixed many thing during testing stage 😄 )

## Background
**Threat Intel Downloader** is a Lambda function triggered by CloudWatch Event periodically to retrieve IOCs ([Indicators of Compromise](https://en.wikipedia.org/wiki/Indicator_of_compromise)) from threat feeds which support RESTful API calls, and it stores IOCs in DynamoDB table where StreamAlert rule processor will refer to while doing threat detection. 

### Supported features
* Retrieve IOCs from threat feeds which support RESTful API calls. And update IOCs periodically, once a day by default. It is scheduled by CloudWatch Event. 
* Store retrieved IOCs in DynamoDB table.
* API credentials are encrypted and store in AWS parameter store. 

### Deploy Threat Intel Downloader
* It is required to initialized StreamAlert before enabling Threat Intel Downloader. Please following [StreamAlert documentation](https://streamalert.readthedocs.io/deployment.html) to initialize StreamAlert. 
* After StreamAlert initialization, you can enable Threat Intel Downloader feature using following command
```
python manage.py threat_intel_downloader enable --interval 'rate(1 day)' --timeout 120 --memory 128
``` 
  * This command will do two things
    1. Generate settings for Threat Intel Downloader lambda function and write them into `conf/lambda.json` file.
    2. Ask you to provide API key and username to fetch IOCs from threat feed which supports RESTful API calls. And store them in AWS Parameter Store with encryption.
```
Please supply API key to retrieve IOCs via API calls. This should be a string of 40 alphanumeric characters.: 0123456789abcdefghijklmnopqrstuvwxyzabcd

Please supply API username to retrieve IOCs via API calls. This should be an email address.: ti-testing@foo.bar
```
  * Deploy Threat Intel Downloader to your infrastructure with following one line command. When the command finishes, you have Threat Intel Downloader Lambda function running once a day and retrieve IOCs from threat feed. 
```
python manage.py lambda deploy --processor threat_intel_downloader
```

## Changes
### Lambda function
* A new Lambda function `Threat Intel Downloader` which is running independently from Rule and Alert processors.
* This lambda function is triggered by CloudWatch Event initially and it will invoke itself if there are more IOCs need to be retrieved when lambda function approaching to timeout. 
* This lambda function will send filtered IOCs to DynamoDB table. StreamAlert rule processors will refer to this table when it is doing threat detection (this feature is coming soon).

### Terraform module
* Add a `Threat Intel Downloader` module to create
  * IAM role
  * Lambda function
  * DynamoDB table
  * CloudWatch Event
  * CloudWatch Log Group

### StreamAlert CLI
* Add new CLI code to support enabling Threat Intel Downloader feature via command line.
* Add new CLI code to generate Threat Intel Downloader configuration and terraform settings.

## Testing
* Unit test was passed
```
./tests/scripts/unit_tests.sh
```
* Tested in AWS test account and it was passed.

jacknagz commented on terraform/modules/tf_threat_intel_downloader/main.tf:
```diff
@@ -0,0 +1,76 @@
+}
+
+// Lambda Alias: Threat Intel Downloader Production
+resource "aws_lambda_alias" "threat_intel_downloader_production" {
```
you can just name this resource `production`

jacknagz commented on terraform/modules/tf_threat_intel_downloader/main.tf:
```diff
@@ -0,0 +1,76 @@
+
+// Lambda Permission: Allow Cloudwatch Scheduled Events to invoke Lambda
+resource "aws_lambda_permission" "allow_cloudwatch_events_invocation" {
+  statement_id  = "CloudwatchEventsInvokeAthenaRefresh"
```
update this SID

jacknagz commented on terraform/modules/tf_threat_intel_downloader/main.tf:
```diff
@@ -0,0 +1,76 @@
+// Cloudwatch Event Rule: Invoke the threat_intel_downloader once a day
+resource "aws_cloudwatch_event_rule" "invoke_threat_intel_downloader" {
+  name        = "invoke_threat_intel_downloader"
+  description = "Invoke the Threat Intel Downloader Lambda function once a day"
```
couldn't it be arbitrarily invoked? maybe just make the description more generic - `Invoke the Threat Intel Downloader Lambda function periodically`

jacknagz commented on terraform/modules/tf_threat_intel_downloader/main.tf:
```diff
@@ -0,0 +1,76 @@
+// Log Retention Policy: lambda function
+resource "aws_cloudwatch_log_group" "threat_intel_downloader" {
+  name              = "/aws/lambda/${var.prefix}_streamalert_threat_intel_downloader"
+  retention_in_days = 60
```
make this a variable

jacknagz commented on terraform/modules/tf_threat_intel_downloader/main.tf:
```diff
@@ -0,0 +1,76 @@
+  retention_in_days = 60
+}
+
+# TODO: check if `aws_cloudwatch_log_metric_filter` needed.
```
leftover comments?

jacknagz commented on terraform/modules/tf_threat_intel_downloader/variables.tf:
```diff
@@ -0,0 +1,72 @@
+
+variable "lambda_handler" {
+  type    = "string"
+  default = "main.handler"
```
this seems invalid considering how we package lambda functions, I'd just nix it

jacknagz commented on terraform/modules/tf_threat_intel_downloader/dynamodb.tf:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  tags {
+    Name = "ThreatIntel"
```
should this be StreamAlert so it's picked up by billing?

chunyong-lin replied:
Hmmm, I will change it to `StreamAlert`. The reason I chose `ThreatIntel` as tag is because we may want to see how much does this standalone lambda function (Threat Intel Downloader) cost. Actually, we can easily get sense since this lambda function mainly use DynamoDB and it is the only feature using DynamoDB in StreamAlert.

ryandeivert replied:
@chunyong-lin you can use multiple tags here if you'd like to keep threatintel as well

jacknagz commented on stream_alert_cli/threat_intel_downloader.py:
```diff
@@ -0,0 +1,101 @@
+    else:
+        LOGGER_CLI.error('Threat Intel Downloader credentials were not saved to parameter store.')
+
+    return saved
```
is anything storing this value?

ryandeivert commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,260 @@
+        """
+        retry, retry_num, continue_invoke = True, self._MAX_RETRY, False
+        intelligence = list()
+        while retry:
```
I think a good amount of this complex logic can be handled using the `backoff` library via a few stacked decorators. Maybe @jacknagz would be willing to offer some pointers here?

chunyong-lin replied:
Thanks! I looked `backoff` library last night, it is very useful. I will include it in my next commit.

ryandeivert commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,260 @@
+                if cred['Name'] == self._PARAMETER_NAME:
+                    try:
+                        decoded_creds = json.loads(cred['Value'])
+                        self.api_user = decoded_creds['api_user']
```
again - only put the items that may result in a `ValueError` in the try block. (move the `self.api_user = decoded_creds['api_user']` and `self.api_key = decoded_creds['api_key']` outside of this)

chunyong-lin commented on manage.py:
```diff
@@ -1157,6 +1156,101 @@ def _add_athena_subparser(subparsers):
+
+Examples:
+
+    manage.py threat_intel_downloader enable \\
```
I want to make enabling Threat Intel Downloader feature as simple as I can. One commandline is easier for to use. And this command is just for generating lambda function settings and writing settings to lambda.json file.

ryandeivert replied:
If we move to this type of pattern, we may as well have a base config that has _all_ possible lambda configs in it and just have an enable/disable flag in them (so we wouldn't be 'adding' a default config to it if it's **enabled** at some point and just would be disabled by default)

chunyong-lin commented on stream_alert_cli/terraform/threat_intel_downloader.py:
```diff
@@ -0,0 +1,57 @@
+    Returns:
+        dict: Athena dict to be marshalled to JSON
+    """
+    # Use the monitoring topic as a dead letter queue
```
I adopted this pattern from StreamAlert App. Using Dead Letter Queue, we can indicate to Lambda that unprocessed events should be sent to SNS topic in our case. Then it will trigger CloudWatch alarm.

ryandeivert replied:
@jacknagz what do you mean by 'how are they handled?' The DLQ SNS topic will receive the messages from the lambda if there is a failure, and will do whatever actions are linked to that sns topic.

chunyong-lin commented on stream_alert_cli/threat_intel_downloader.py:
```diff
@@ -0,0 +1,101 @@
+        options (namedtuple): The parsed args passed from the CLI
+        config (CLIConfig): Loaded StreamAlert CLI
+    """
+    def _validate_options(options):
```
Yes, it is enforced at the parser level. This method is imported and used in `stream_alert_cli/runner.py` where doing command parsing.

chunyong-lin commented on manage.py:
```diff
@@ -1157,6 +1156,114 @@ def _add_athena_subparser(subparsers):
+        help=ARGPARSE_SUPPRESS
+    )
+
+    ti_downloader_parser.add_argument(
```
Thanks for the suggestion. The helper functions are very useful!

chunyong-lin commented on manage.py:
```diff
@@ -1157,6 +1156,114 @@ def _add_athena_subparser(subparsers):
+        default=['crowdstrike', '@airbnb.com']
+    )
+
+    ti_downloader_parser.add_argument(
```
Unfortunately, the name of IOC types will be different from different threat feeds. We can not fix IOC types.

ryandeivert commented on stream_alert_cli/config.py:
```diff
@@ -125,6 +125,8 @@ def set_prefix(self, prefix):
             'rule_processor_config']['source_bucket'].replace('PREFIX_GOES_HERE', prefix)
         self.config['lambda']['stream_alert_apps_config']['source_bucket'] = self.config['lambda'][
             'stream_alert_apps_config']['source_bucket'].replace('PREFIX_GOES_HERE', prefix)
+        self.config['lambda']['threat_intel_downloader_config']['source_bucket'] = self.config['lambda'][
```
@jacknagz that replace only happens if it exists, so what's wrong with it being here?

ryandeivert commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,260 @@
+                        }
+                    ]
+        """
+        def _epoch_time(time_str):
```
making this is `@staticmethod` instead of a nested function would make testing this much easier. I'd avoid nested functions if possible. Currently there is no way to reference this function from outside of its containing function, so testing it properly becoming inherently difficult.

ryandeivert commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,260 @@
+            """
+
+            if not time_str:
+                return int((datetime.now() + timedelta(90) - datetime(1970, 1, 1)).total_seconds())
```
I think the preferred way for the latter part of this (`datetime(1970, 1, 1)`) would be `datetime.utcfromtimestamp(0)`.

Can you also add the `days=90` keyword arg to the `timedelta(90)` call, since timedelta can also take `seconds=`, `minutes=` - it just helps with readability without having to reference the docstring above.

ryandeivert commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,260 @@
+                utc_time = datetime.strptime(time_str, "%Y-%m-%dT%H:%M:%S.%fZ")
+                return int((utc_time - datetime(1970, 1, 1)).total_seconds())
+            except ValueError:
+                LOGGER.error('Can not convert expiration date \'%s\' to epoch time',
```
`Can not` --> `Cannot`

chunyong-lin commented on stream_alert/threat_intel_downloader/main.py:
```diff
@@ -0,0 +1,94 @@
+        (dict): Configuration for Threat Intel Downloader
+
+    Raises:
+        ConfigError: For invalid or missing configuration files.
```
Good catch! Thanks!

chunyong-lin commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,260 @@
+            return None, None, False
+
+        next_url = None
+        if 'next_url' not in event:
```
Your 🚀  
Your understanding is correct. The `event` may contain `next_url` key but with empty value. It is just special handling in following lines.

chunyong-lin commented:
@jacknagz @ryandeivert PTAL
I have addressed your comments. Thank you so much, they are really helpful 😄 
Mainly changes are,
* Combine methods to generate lambda settings for `Athena` and `Threat Intel Downloader`
* Use existing `MockSSMClient` class during testing
* Promote helpers for validating arguments to be globally accessed
* Use `backoff` library to handle `requests` retries

I noticed, the coverage is dropped by 1.2%. I will add more test cases.

ryandeivert commented on stream_alert_cli/helpers.py:
```diff
@@ -30,10 +30,56 @@
+class ParseError(Exception):
+    """Custom expection rasied when argument parser has error"""
+
+def validate_scheduled_interval(val):
```
As discussed offline, can we actually do away with the helpers and just duplicate this code (sigh) in the arg parsers? I don't love that this will actually raise an exception and print the traceback, instead of acting like the normal argparse error that prints the usage and the error (without printing a traceback). I'm sorry :(

chunyong-lin replied:
:( Yeah, I will make change. I am agree that it is ugly to see the traceback when there is argument parse error. So we have to make a compromise that to use clean build-in argument error handling (have to duplicate code) over custom exception which will print out traceback.

ryandeivert commented on tests/unit/app_integrations/test_helpers.py:
```diff
@@ -24,8 +24,9 @@
-        self._parameters = dict()
+    def __init__(self, suppress_params=False, app_type='', parameters=None):
+        #self._parameters = dict()
+        self._parameters = parameters if parameters else dict()
```
you can just do `self._parameters = parameters or dict()`

ryandeivert commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,273 @@
+                raise ThreatStreamRequestsError('Unknown status code {}, '
+                                                'do not retry.'.format(https_req.status_code)
+                                               )
+        except (requests.exceptions.Timeout,
```
hey @chunyong-lin, I think when using `backoff`, you shouldn't wrap the code in a try/except block like this. You should be able to get rid of this and achieve what you're trying to do.

chunyong-lin replied:
Ha, good catch! I will remove `try..except`.

ryandeivert commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,273 @@
+
+        if not time_str:
+            return int((datetime.now()
+                        + timedelta(days)
```
this is interesting formatting. pylint doesn't complain about this 🤔 ?

I'd suggest putting the `+` and `-` on the preceding line as such:
```
return int((datetime.now() +
            timedelta(days) -
            datetime.utcfromtimestamp(0)).total_seconds())
```

chunyong-lin replied:
[PEP8](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator) recommends to break lines before operants, but actually before and after operants are both okay if we have style consistently.

ryandeivert replied:
Weird 🤔

jacknagz commented on terraform/modules/tf_threat_intel_downloader/iam.tf:
```diff
@@ -0,0 +1,144 @@
+    ]
+
+    resources = [
+      "arn:aws:lambda:${var.region}:${var.account_id}:function:${var.prefix}_streamalert_threat_intel_downloader",
```
I would feed this in via variables vs interpolation

jacknagz replied:
you can use outputs from the main `tf_stream_alert` module

chunyong-lin replied:
I am following the same style to interpolate resources across streamalert terraform modules, example in [streamalert iam tf file](https://github.com/airbnb/streamalert/blob/master/terraform/modules/tf_stream_alert/iam.tf#L48)

jacknagz commented on terraform/modules/tf_threat_intel_downloader/iam.tf:
```diff
@@ -0,0 +1,144 @@
+    ]
+
+    resources = [
+      "arn:aws:dynamodb:${var.region}:${var.account_id}:table/${var.prefix}_streamalert_threat_intel_downloader",
```
can this be interpolated?

chunyong-lin replied:
Can you have more content? Do you mean interpolated in different way?

jacknagz commented on terraform/modules/tf_threat_intel_downloader/iam.tf:
```diff
@@ -0,0 +1,144 @@
+
+// IAM role policy: Allow lambda function to read from parameter store
+resource "aws_iam_role_policy" "read_api_creds_from_ssm" {
+  name   = "read_ssm"
```
I'd make this camel case, like the others:

`SSMGetThreatIntelParms`

chunyong-lin commented on stream_alert/threat_intel_downloader/threat_stream.py:
```diff
@@ -0,0 +1,273 @@
+                             requests.exceptions.ChunkedEncodingError,
+                             ThreatStreamRequestsError)
+    @backoff.on_exception(backoff.constant,
+                          exceptions_to_backoff,
```
I thought about using `expo`, but I don't want to add too much timeout. I can change to `expo` if constant doesn't work well. Using constant with 3 times retries, the max timeout will be 3 seconds. If use `expo`, it will take `1+2+4 = 7` seconds.

jacknagz replied:
expo is actually not linear like that, but sounds good!

jacknagz replied:
there's jitter/randomization involved

chunyong-lin commented on stream_alert_cli/config.py:
```diff
@@ -431,6 +438,55 @@ def add_app_integration(self, app_info):
+        Returns:
+            (bool): Return True if writing settings of Lambda function successfully.
+        """
+        def _generate_default_ti_downloader_configure():
```
It is weird that I use function here 😄  Don't know what I was doing 😂 
Will change to  variable.

chunyong-lin commented:
@jacknagz PTAL. I addressed some of your comments, thanks! Also, I may need more more context about two of your comments. Please see above.
